src/hotspot/share/runtime/mutexLocker.cpp

```diff
@@ -372,3 +372,26 @@ void print_owned_locks_on_error(outputStream* st) {
   }
   if (none) st->print_cr("None");
 }
+
+void unlock_locks_owned_by(Thread* thread) {
+  assert(thread != NULL, "can't be owned by NULL");
+  if (thread->is_Watcher_thread()) {
+    // need WatcherThread as a safeguard against potential deadlocks
+    return;
+  }
+
+#ifdef ASSERT
+  Mutex* owned_lock = thread->owned_locks();
+  while (owned_lock != NULL) {
+    Mutex* next = owned_lock->next();
+    owned_lock->unlock();
+    owned_lock = next;
+  }
+#endif // ASSERT
+
+  for (int i = 0; i < _num_mutex; i++) {
+     if (_mutex_array[i]->owner() == thread) {
+       _mutex_array[i]->unlock();
+     }
+  }
```

Comment:
Surely we don't need this in a debug build as we already unlocked all owned locks?

Reply:
This logic is for release build. 

Yes, we don't need them in debug build. We should have released all owning mutexes above. 
It's no-op in debug build because owner() should be NULL.  `thread` isn't  NULL in this function. 

Here is current bookkeeping. From a thread, we can't find all owning mutexes. @pchilano said we can try _mutex_array in release build. I think the idea is that we are supposed to capture failure in debug build. 

| class       | debug                     | release      |   |
|-------------|---------------------------|--------------|---|
| Thread      | _owned_locks (listedlist) | N/A          |   |
| Mutex       | _owner                    | _owner       |   |
| MutexLocker | _mutex_array              | _mutex_array |   |

Shall we fix _mutex_array in release? I think we can change it to GrowableArray and keep track all mutexes in runtime.

Reply:
My point is that the code to process the `_mutex_array` should be in a `#else` so that we don't build it or use for debug builds. So a debug build uses `owned_locks()` while a release build uses `_mutex_array`.

Reply:
Sorry if I missed part of the conversation, but why two different implementations? Why not use _mutex_array always? I dislike having different code in debug and release builds.

Reply:
https://bugs.openjdk.java.net/browse/JDK-8274794 I also think we should have _owned_locks available in product.